py/server/deephaven/barrage.py

```diff
@@ -0,0 +1,176 @@
+#
+# Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+#
+""" This module defines the BarrageSession wrapper class and provides a factory function to create an instance of it
+ for accessing resources on remote Deephaven servers."""
+
+from __future__ import annotations
+import threading
+from typing import Dict
+
+import jpy
+
+from deephaven import DHError, uri
+from deephaven._wrapper import JObjectWrapper
+from deephaven.table import Table
+
+_JURI = jpy.get_type("java.net.URI")
+_JClientConfig = jpy.get_type("io.deephaven.client.impl.ClientConfig")
+_JSSLConfig = jpy.get_type("io.deephaven.ssl.config.SSLConfig")
+_JSessionImplConfig = jpy.get_type("io.deephaven.client.impl.SessionImplConfig")
+_JTrustCustom = jpy.get_type("io.deephaven.ssl.config.TrustCustom")
+_JDeephavenTarget = jpy.get_type("io.deephaven.uri.DeephavenTarget")
+_JBarrageSession = jpy.get_type("io.deephaven.client.impl.BarrageSession")
+_JSharedId = jpy.get_type("io.deephaven.client.impl.SharedId")
+_JBarrageTableResolver = jpy.get_type("io.deephaven.server.uri.BarrageTableResolver")
+_JChannelHelper = jpy.get_type("io.deephaven.client.impl.ChannelHelper")
+_JDeephavenChannelImpl = jpy.get_type("io.deephaven.proto.DeephavenChannelImpl")
+_JSessionImpl = jpy.get_type("io.deephaven.client.impl.SessionImpl")
+_JExecutors = jpy.get_type("java.util.concurrent.Executors")
+_JRootAllocator = jpy.get_type("org.apache.arrow.memory.RootAllocator")
+
+_session_cache: Dict[str, BarrageSession] = {}  # use WeakValueDictionary to avoid memory leak?
+_remote_session_lock = threading.Lock()
+
+
+def barrage_session(host: str,
+                    port: int = 10000,
+                    auth_type: str = "Anonymous",
+                    auth_token: str = "",
+                    use_tls: bool = False,
+                    tls_root_certs: bytes = None,
+                    ) -> BarrageSession:
```

Comment:
I think we should purposefully limit this signature ATM, either to `barrage_session(dh_target: str)` or `barrage_session(host, port, use_tls)` (I prefer the former, as it's more like URI, `barrage_session("dh://somehost")`; I know the intention is to make it look like pydeephaven, but I don't think that's actually what we want to support long term. In a lot of ways, this is similar to the BarrageTableResolver / URI API, where we have a more general issues that we want to eventually solve w/ some sort of context that provides authentication and TLS info. (https://github.com/deephaven/deephaven-core/issues/5383) I worry that by trying to provide all this data manually we are diverging from what could be a more complete java server side solution. A limited API here still allows users the workaround of creating their own BarrageSession (like DHE will do), and I think tickets asking for authentication support should really be pushing us to offer more general server side solution for it.

Reply:
For the immediate problem, we need a convenient way for a worker using both `deephaven` and `pydeephaven` to create the session.  If the above is changed to differ from `pydeephaven`, this detail needs to be considered.

Reply:
No, we don't. That's a "nice to have" feature for community-only use cases, rather than a requirement for the initial version of this feature. It's necessarily the case that we won't be able to own the `BarrageSession` creation in the target use case.

Reply:
I am still in favor of keep the two signatures as similar as possible. It is purely from my own experience of writing the client and server code in the same script. Unless we don't want users to use pydeephaven in the server and expand the new barrage module to do pretty much what pydeephaven can do, then having very similar signatures is better.

Reply:
Ideally, when a user is creating pydeephaven on the server, they'll be able to hook into a potentially easier signature and pick up TLS and authentication by default in the future, much like how we want URIs to work:

```python
# will pick up default TLS, auth, etc from the server (TBD)
foo = uri.resolve("dh://remotehost/scope/foo")
```

```python
# will pick up default TLS, auth, etc from the server (TBD)
client = pydeephaven("dh://remotehost")
foo_remote = client.open_table("foo")
...
```

Reply:
It seems like @jmao-denver is advocating for an API that works now while @devinrsmith is advocating for the API of the future.  Is enough of the future implemented to make something work, or are we really limited by what we have today and should stick with what is currently implemented for now?